tensorflow/core/kernels/sparse_tensor_dense_matmul_op_gpu.cu.cc

```diff
@@ -71,62 +101,111 @@ __global__ void SparseTensorDenseMatMulKernel(
+}
+
+} // namespace
+
 template <typename T, typename Tindices, bool ADJ_A, bool ADJ_B>
 struct SparseTensorDenseMatMulFunctor<GPUDevice, T, Tindices, ADJ_A, ADJ_B> {
+
   static EIGEN_ALWAYS_INLINE Status
   Compute(OpKernelContext* ctx, typename TTypes<T>::Matrix out,
           typename TTypes<Tindices>::ConstMatrix a_indices,
           typename TTypes<T>::ConstVec a_values,
           typename TTypes<T>::ConstMatrix b) {
-    int nnz = a_values.size();
-    // out = A * B, A is [m x n] and B is [n x p], out is [m x p]
-    int m = out.dimension(0);
-    int p = out.dimension(1);
-    int b_rows = b.dimension(0);
-    int b_cols = b.dimension(1);
-
-    const GPUDevice& d = ctx->eigen_device<GPUDevice>();
-    using Tsum = typename SumType<T>::type;
-    Tsum* maybe_temp_out_data = nullptr;
-    Tensor temp_out_t;
-    bool sum_type_is_different = !std::is_same<T, Tsum>::value;
-    if (sum_type_is_different) {
-      TF_RETURN_IF_ERROR(ctx->allocate_temp(
-          DataTypeToEnum<Tsum>::value,
-          TensorShape({out.dimension(0), out.dimension(1)}), &temp_out_t));
-      auto temp_out = temp_out_t.matrix<Tsum>();
-      maybe_temp_out_data = temp_out.data();
-      temp_out.device(d) = temp_out.constant(Tsum(0));
-    } else {
-      // Note: The reinterpret cast is only required to avoid a compilation
-      // error; it is only used if Tsum == T.
-      maybe_temp_out_data = reinterpret_cast<Tsum*>(out.data());
-      out.device(d) = out.constant(T(0));
-    }
 
-    // TODO(ebrevdo): Should this be alpha * nnz instead of
-    // out.size()?  Perhaps p * nnz ?
-    GpuLaunchConfig config = GetGpuLaunchConfig(p * nnz, d);
-
-    TF_CHECK_OK(GpuLaunchKernel(
-        SparseTensorDenseMatMulKernel<T, Tsum, Tindices, ADJ_A, ADJ_B>,
-        config.block_count, config.thread_per_block, 0, d.stream(), nnz, m,
-        b_rows, b_cols, p, a_indices.data(), a_values.data(), b.data(),
-        maybe_temp_out_data));
-
-    if (sum_type_is_different) {
-      out.device(d) = temp_out_t.matrix<Tsum>().template cast<T>();
+    if (RequireDeterminism()) {
+      if (std::is_same<T, double>::value ||
+          std::is_same<T, complex128>::value) {
+        return errors::Unimplemented(
+            "No deterministic GPU implementation of sparse_dense_matmul "
+            "available for data of type tf.float64 or tf.complex128");
+      }
+      using Tsum = typename SumType<T>::type_for_determinism;
+      ComputeWithPossiblyHigherPrecisionAccumulator <Tsum>(ctx, out, a_indices,
+                                                 a_values, b);
+    } else {
+      using Tsum = typename SumType<T>::type;
+      ComputeWithPossiblyHigherPrecisionAccumulator <Tsum>(ctx, out, a_indices,
+                                                 a_values, b);
     }
-
     return Status::OK();
   }
+
+  private:
+    template <typename Tsum>
+    static EIGEN_ALWAYS_INLINE Status
+    ComputeWithPossiblyHigherPrecisionAccumulator (
```

Comment:
There should be no space between the function name and the opening parenthesis. See [here](https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions).

Reply:
Note to self: addressed.